File: Python/lc_207_course_schedule.py
# 207. Course Schedule
# Time: 2E + V
# Space: E + 2V
# Topological Sort
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        import collections

        dep_map = collections.defaultdict(list)
        dep_count = collections.defaultdict(int)

        for dep, course in prerequisites:
            dep_map[course].append(dep)
            dep_count[dep]+=1

        indies = [course for course in range(numCourses) if course not in dep_count]

        indie_count = 0

        while indies:

            cur_course = indies.pop(0)

            for course in dep_map[cur_course]:
                dep_count[course]-=1
                if dep_count[course]==0:
                    del dep_count[course]
                    indies.append(course)

        if len(dep_count)>0:
            return False
        return True

# Time: O(V+E) vertices + edges
# Space: O(V+E)
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        from collections import defaultdict

        prereq_map = defaultdict(list)

        for course, prereq in prerequisites:
            prereq_map[course].append(prereq)

        visited = ["unvisited" for _ in range(numCourses)]

        for cur_course in range(numCourses):

            if not self.dfs(cur_course, visited, prereq_map):
                return False

        return True


    def dfs(self, cur_course, visited, prereq_map):

        if visited[cur_course]=="under_progress":
            # We hit a course that hasn't been completed, hence there's a loop, i.e a course is dependant on itself for completion
            return False
        elif visited[cur_course]=="done":
            return True
        else:
            visited[cur_course] = "under_progress"

        for prereq in prereq_map[cur_course]:
            if not self.dfs(prereq, visited, prereq_map):
                return False

        visited[cur_course]="done"
        return True

File: Go/lc207CourseSchedule.go
package Go

// Topological Sort solution
// Time Complexity: O(V+2E)
func canFinish(numCourses int, prerequisites [][]int) bool {

	deps := make(map[int][]int)
	numDeps := make(map[int]int)
	for _, val := range prerequisites {
		course := val[0]
		dep := val[1]

		if _, ok := deps[dep]; !ok {
			deps[dep] = []int{}
		}
		deps[dep] = append(deps[dep], course)
		numDeps[course]++
	}

	topSortStack := []int{}

	for i := 0; i < numCourses; i++ {
		if numDeps[i] == 0 {
			topSortStack = append(topSortStack, i)
		}
	}

	for len(topSortStack) > 0 {
		curCourse := topSortStack[len(topSortStack)-1]
		topSortStack = topSortStack[:len(topSortStack)-1]

		for _, course := range deps[curCourse] {
			numDeps[course] = numDeps[course] - 1
			if numDeps[course] == 0 {
				delete(numDeps, course)
				topSortStack = append(topSortStack, course)
			}
		}
	}

	if len(numDeps) > 0 {
		return false
	}
	return true
}

// Identifying cyclic dependancy solution
// Time Complexity: O(V+E)
func canFinishTopological(numCourses int, prerequisites [][]int) bool {

	deps := make(map[int][]int)

	for _, v := range prerequisites {

		course, dep := v[0], v[1]

		if _, ok := deps[course]; !ok {
			deps[course] = []int{}
		}
		deps[course] = append(deps[course], dep)
	}

	for i := 0; i < numCourses; i++ {

		if _, ok := deps[i]; !ok {
			continue
		}

		for _, val := range deps[i] {

			tempMap := make(map[int]string)

			if !traverse(deps, val, tempMap) {
				return false
			}
		}
	}

	return true
}

func traverse(deps map[int][]int, cur int, trackMap map[int]string) bool {

	if val, ok := trackMap[cur]; ok {
		if val == "in_progress" {
			return false
		}
		return true
	}

	trackMap[cur] = "in_progress"

	for _, v := range deps[cur] {

		if !traverse(deps, v, trackMap) {
			return false
		}
	}

	trackMap[cur] = "done"

	return true
}
